library/Freckle/App/Http/Cache/Memcached.hs

```diff
@@ -0,0 +1,144 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedRecordDot #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Freckle.App.Http.Cache.Memcached
+  ( memcachedHttpCacheSettings
+  , memcachedHttpCodec
+  , memcachedHttpCache
+  ) where
+
+import Freckle.App.Prelude
+
+import Blammo.Logging (MonadLogger, logWarnNS)
+import Codec.Serialise (Serialise (..), deserialiseOrFail, serialise)
+import qualified Data.ByteString.Lazy as BSL
+import Data.CaseInsensitive (CI)
+import qualified Data.CaseInsensitive as CI
+import Data.String (IsString (..))
+import Database.Memcache.Types (Value)
+import Freckle.App.Http.Cache
+import Freckle.App.Memcached
+import qualified Freckle.App.Memcached.Client as Memcached
+import Freckle.App.OpenTelemetry (MonadTracer)
+import Network.HTTP.Client (Cookie, CookieJar, Request)
+import qualified Network.HTTP.Client.Internal as HTTP
+import Network.HTTP.Types.Header (ResponseHeaders)
+import Network.HTTP.Types.Status (Status)
+import Network.HTTP.Types.Version (HttpVersion (..))
+
+memcachedHttpCacheSettings
+  :: ( MonadUnliftIO m
+     , MonadLogger m
+     , MonadTracer m
+     , MonadReader env m
+     , HasMemcachedClient env
+     , Serialise body
+     )
+  => CacheTTL
+  -- ^ Default TTL, used when @max-age@ is not present
+  -> HttpCacheSettings m body Value
+memcachedHttpCacheSettings defaultTTL =
+  HttpCacheSettings
+    { shared = False
+    , cacheable = const True
+    , defaultTTL
+    , getCurrentTime = liftIO getCurrentTime
+    , logWarn = logWarnNS "http.cache" . fromString
+    , codec = memcachedHttpCodec
+    , cache = memcachedHttpCache
+    }
+
+memcachedHttpCodec :: Serialise body => HttpCacheCodec body Value
+memcachedHttpCodec =
+  HttpCacheCodec
+    { serialise = BSL.toStrict . serialise . fromResponse
+    , deserialise = \req ->
+        bimap show (toResponse req)
+          . deserialiseOrFail
+          . BSL.fromStrict
+    }
+
+memcachedHttpCache
+  :: ( MonadUnliftIO m
+     , MonadTracer m
+     , MonadReader env m
+     , HasMemcachedClient env
+     )
+  => HttpCache m Value
+memcachedHttpCache =
+  HttpCache
+    { get = try . Memcached.get
+    , set = \k v -> try $ Memcached.set k v 0
+    , evict = try . Memcached.delete
+    }
+
+data CResponse body = CResponse
+  { cresponseStatus :: Status
+  , cresponseVersion :: HttpVersion
+  , cresponseHeaders :: ResponseHeaders
+  , cresponseBody :: body
+  , cresponseCookieJar :: CookieJar
+  , cresponseEarlyHints :: ResponseHeaders
+  , cinserted :: UTCTime
+  , cttl :: CacheTTL
+  }
+  deriving stock (Generic)
+  deriving anyclass (Serialise)
+
+{- FOURMOLU_DISABLE -}
+-- Fourmolu has trouble with this bit of CPP
+
+toResponse :: Request -> CResponse body -> CachedResponse body
```

Comment:
I am having trouble remembering the distinction between `CResponse` and `CachedResponse`, I don't know if I need different names or just comments or what.

Reply:
It is annoying. `CachedResponse` is the value that goes into cache (any cache). And `CResponse` is just something we have to convert that to in order to give it a `Serialise` instance.

Things would be a lot easier if I could just give `Response body` an orphan instance, and then derive `Serialise` for `CachedResponse`. But the trouble is `responseOriginalRequest` -- get `Serialise` on `Request` looks impossible (it has a few functional fields), so we simply can't serialise an actual `Response`. That's why `CResponse` omits it and then is given in back again when deserialising.

Reply:
Would `s/CResponse/SerialisableResponse/` work?

Reply:
If the purpose of what is now called `CResponse` is its `Serialise` instance, then yeah I think "serialise" in the name would be a good mnenonic. A comment that says "This type defines how to serialize a `CachedResponse`" would do a lot too.